pkg/controller/remotemachineset/gcpactuator.go

```diff
@@ -142,3 +168,126 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
+				ownerRef.Kind == "MachinePool" &&
+				ownerRef.Name == pool.Name {
+				logger.Info("machine pool already has lease: %s", l.Name)
+				return rune(l.Name[len(l.Name)-1]), nil
+			}
+		}
+	}
+
+	// Pool does not have a lease yet, lookup all currently available lease chars
+	availLeaseChars, err := a.findAvailableLeaseChars(cd)
+	if err != nil {
+		return 'X', err
+	}
+	logger.Debugf("availLeaseChars: %v", availLeaseChars)
+
+	// If the pool.Spec.Name == "worker", we want to preseve this MachinePool's "w" character that
+	// the installer would have selected so we do not cycle all worker nodes. In theory, multiple
+	// MachinePools could specify this in their spec.Name, in which case we'll start erroring
+	// because the first one already grabbed "w".
+	var leaseRune rune
+	if pool.Spec.Name == "worker" {
+		for _, r := range availLeaseChars {
+			if r == rune('w') {
+				leaseRune = 'w'
+				logger.Debug("selecting lease char 'w' for original worker pool")
+			}
+		}
+	} else {
+		// Choose a random entry in the available chars to limit collisions while processing
+		// multiple machine pools at the same time. In this case the subsequent attempts to create
+		// the lease will fail due to name collisions and re-reconcile.
+		logger.Debug("selecting random lease char from available")
+		rand.Seed(time.Now().UnixNano())
+		leaseRune = availLeaseChars[rand.Intn(len(availLeaseChars))]
+	}
+	logger.Debugf("selected lease char: %s", string(leaseRune))
+
+	leaseName := fmt.Sprintf("%s-%s", cd.Spec.ClusterMetadata.InfraID, string(leaseRune))
+	// Attempt to claim the lease:
+	l := &hivev1.MachinePoolNameLease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      leaseName,
+			Namespace: pool.Namespace,
+		},
+	}
+	err = a.client.Create(context.TODO(), l)
+	if err != nil {
+		return 'X', err
+	}
+	logger.WithField("lease", leaseName).Infof("created lease")
+
+	return leaseRune, nil
+}
+
+func stringToRuneSet(s string) map[rune]bool {
+	chars := make(map[rune]bool)
+	for _, char := range s {
+		chars[char] = true
+	}
+	return chars
+}
+
+func (a *GCPActuator) findAvailableLeaseChars(cd *hivev1.ClusterDeployment) ([]rune, error) {
+	availChars := stringToRuneSet(validLeaseChars)
+
+	leases := &hivev1.MachinePoolNameLeaseList{}
+	err := a.client.List(context.TODO(), leases, client.UseListOptions(&client.ListOptions{Namespace: cd.Namespace}))
+	if err != nil {
+		return []rune{}, err
+	}
+
+	for _, lease := range leases.Items {
+		if len(lease.OwnerReferences) != 1 {
+			return []rune{}, fmt.Errorf("MachinePoolNameLease %s has unexpected number of owner refs: %d", lease.Name, len(lease.OwnerReferences))
+		}
+
+		if lease.OwnerReferences[0].APIVersion != "hive.openshift.io/v1" || lease.OwnerReferences[0].Kind != "MachinePool" {
+			return []rune{}, fmt.Errorf("MachinePoolNameLease %s has unexpected owner ref: %v", lease.Name, lease.OwnerReferences[0])
+		}
+
+		// TODO: should we store the cluster deployment infra ID on the lease to make this unnecessary?
```

Comment:
Yes. It should be stored in a label, although I don't follow why the infra ID is necessary and not just the clusterdeployment name.

But, we still need to check that the name of the lease is in the correct format. We are relying on that name to provide the uniqueness of the lease, so it **must** be correct.

Reply:
Adding a check that everything but the last char matches infraid- which should cover us in all cases. (and a test)